### PR TITLE
Improve the performance of bevformer

### PR DESCRIPTION
**Pull request by Cathy0908**
**训练性能：**

​	**单机单卡，fp16**:

|                    | Official吞吐 | EasyCV吞吐 | 性能提升 | 环境                              |
| ------------------ | ------------ | ---------- | -------- | --------------------------------- |
| **bevformer-base** | 0.727        | 0.8        | 10%      | A100, batch_size=5, num_worker=8  |
| **bevformer-tiny** | 3.55         | 9.84       | 177%     | A100, batch_size=32, num_worker=8 |

**推理性能**

​	**fp32**  **(精度已对齐)**

​    **性能提升：42%**

|      | Name          | Backend  | Median(FPS) | Mean(FPS) | Median(ms) | Mean(ms) | 99th_p   | std_dev     |
| ---- | ------------- | -------- | ----------- | --------- | ---------- | -------- | -------- | ----------- |
| 0    | easycv        | TensorRT | 3.68697     | 3.68651   | 0.271226   | 0.271259 | 0.272165 | 0.000197132 |
| 1    | easycv script | TensorRT | 3.8131      | 3.79859   | 0.262254   | 0.26337  | 0.295066 | 0.00577846  |
| 2    | blade         | TensorRT | 5.40248     | 5.23383   | 0.1851     | 0.192212 | 0.261654 | 0.0163032   |

### CODE REVIEW

File: easycv/models/detection3d/detectors/bevformer/attentions/spatial_cross_attention.py [hunk 1]
```diff
@@ -108,9 +110,12 @@ def forward(self,
         if value is None:
             value = key
 
-        if residual is None:
-            inp_residual = query
-            slots = torch.zeros_like(query)
+        # if residual is None:
```
Comment (wenmengzhou) [inline easycv/models/detection3d/detectors/bevformer/attentions/spatial_cross_attention.py:113]: assert residula is None

File: configs/detection3d/bevformer/bevformer_tiny_r50_nuscenes.py [hunk 2]
```diff
@@ -179,11 +180,12 @@
 data_root = 'data/nuScenes/nuscenes-v1.0/'
 
 train_pipeline = [
+    dict(type='RandomScaleImageMultiViewImage', scales=[0.5]),
     dict(type='PhotoMetricDistortionMultiViewImage'),
     dict(type='ObjectRangeFilter', point_cloud_range=point_cloud_range),
     dict(type='ObjectNameFilter', classes=CLASSES),
     dict(type='NormalizeMultiviewImage', **img_norm_cfg),
-    dict(type='RandomScaleImageMultiViewImage', scales=[0.5]),
+    # dict(type='RandomScaleImageMultiViewImage', scales=[0.5]),
```
Comment (wenmengzhou) [inline configs/detection3d/bevformer/bevformer_tiny_r50_nuscenes.py:188]: does this affect the final NDS?

File: easycv/models/detection3d/detectors/bevformer/bevformer_head.py [hunk 3]
```diff
@@ -309,20 +307,47 @@ def forward(self, mlvl_feats, img_metas, prev_bev=None, only_bev=False):
...
+            tmp_1_2_new = (
+                tmp_1_2 * (self.pc_range[4] - self.pc_range[1]) +
                 self.pc_range[1])
-            tmp[..., 4:5] = (
-                tmp[..., 4:5] * (self.pc_range[5] - self.pc_range[2]) +
+            tmp_4_5_new = (
+                tmp_4_5_add_reference * (self.pc_range[5] - self.pc_range[2]) +
                 self.pc_range[2])
 
+            tmp_2_4 = tmp[..., 2:4]
```
Comment (wenmengzhou) [inline easycv/models/detection3d/detectors/bevformer/bevformer_head.py:345]: when in jit tracing or jit script, use  the changed version , whle in training use inplace operation to get better performance？

File: easycv/models/detection3d/detectors/bevformer/transformer.py [hunk 4]
```diff
@@ -249,43 +267,42 @@ def forward(self,
...
             elif layer == 'cross_attn':
                 query = self.attentions[attn_index](
-                    query,
-                    key,
-                    value,
-                    identity if self.pre_norm else None,
+                    query=query,
+                    key=key,
+                    value=value,
+                    # residual=identity if self.pre_norm else None,
```
Comment (wenmengzhou) [inline easycv/models/detection3d/detectors/bevformer/transformer.py:297]: add condition check

File: easycv/models/detection3d/detectors/bevformer/transformer.py [hunk 5]
```diff
@@ -355,13 +371,26 @@ def forward(self,
...
 
-                new_reference_points = torch.zeros_like(reference_points)
-                new_reference_points[..., :2] = tmp[..., :2] + inverse_sigmoid(
-                    reference_points[..., :2], eps=1e-5)
-                new_reference_points[...,
-                                     2:3] = tmp[..., 4:5] + inverse_sigmoid(
-                                         reference_points[..., 2:3], eps=1e-5)
+                new_reference_points = torch.zeros_like(
+                    reference_points)  # torch.Size([1, 900, 3])
+                # new_reference_points[..., :2] = tmp[..., :2] + inverse_sigmoid(
```
Comment (wenmengzhou) [inline easycv/models/detection3d/detectors/bevformer/transformer.py:376]: same as before

File: easycv/predictors/bevformer_predictor.py [hunk 6]
```diff
@@ -40,8 +45,11 @@ def __init__(self,
                  box_type_3d='LiDAR',
                  use_camera=True,
                  score_threshold=0.1,
+                 model_type=None,
                  *arg,
                  **kwargs):
+        self.model_type = model_type
+        self.is_jit_model = self.model_type in ['jit', 'blade']
```
Comment (wenmengzhou) [inline easycv/predictors/bevformer_predictor.py:52]: judge model_type according to extension？  just as YoloXPredictor

File: easycv/runner/ev_runner.py [hunk 7]
```diff
@@ -94,7 +94,7 @@ def train(self, data_loader, **kwargs):
         self.data_loader = data_loader
         self._max_iters = self._max_epochs * len(self.data_loader)
         self.call_hook('before_train_epoch')
-        time.sleep(2)  # Prevent possible deadlock during epoch transition
+        # time.sleep(2)  # Prevent possible deadlock during epoch transition
```
Comment (wenmengzhou) [inline easycv/runner/ev_runner.py:97]: delete these two lines

File: easycv/thirdparty/deformable_attention/src/vision.cpp [hunk 8]
```diff
@@ -14,8 +14,21 @@
...
+    const at::Tensor &value, 
+    const at::Tensor &spatial_shapes,
+    const at::Tensor &level_start_index,
+    const at::Tensor &sampling_loc,
+    const at::Tensor &attn_weight,
+    const int64_t im2col_step) {
+  return ms_deform_attn_forward(value, spatial_shapes, level_start_index, 
+    sampling_loc, attn_weight, im2col_step);
+}
+static auto registry = torch::RegisterOperators().op("custom::ms_deform_attn", &ms_deform_attn);
```
Comment (wenmengzhou) [inline easycv/thirdparty/deformable_attention/src/vision.cpp:34]: it is better to add a compiling script

File: tests/file/test_image.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+class LoadImageTest(unittest.TestCase):
+    img_path = os.path.join(TEST_IMAGES_DIR, '000000289059.jpg')
+    img_url = 'http://pai-vision-data-hz.oss-cn-zhangjiakou.aliyuncs.com/unittest/local_backup/easycv_nfs/data/test_images/000000289059.jpg'
+
+    def setUp(self):
+        print(('Testing %s.%s' % (type(self).__name__, self._testMethodName)))
+
+    def test_backend_pillow(self):
+        img = load_image(
+            self.img_path, mode='BGR', dtype=np.float32, backend='pillow')
```
Comment (wenmengzhou) [inline tests/file/test_image.py:21]: why mode is BGR and the value is the same as RGB using turbojpeg and cv2

File: configs/detection3d/bevformer/bevformer_tiny_r50_nuscenes.py (hunk 2 above)
Comment (Cathy0908) [reply configs/detection3d/bevformer/bevformer_tiny_r50_nuscenes.py:188]: 精度没测，但是从pipeline的顺序和功能上看没影响

File: tests/file/test_image.py (hunk 9 above)
Comment (Cathy0908) [reply tests/file/test_image.py:21]: 不一样啊，一个是[145, 92, 59]，一个是[59, 92, 145]

File: easycv/models/detection3d/detectors/bevformer/transformer.py (hunk 4 above)
Comment (Cathy0908) [reply easycv/models/detection3d/detectors/bevformer/transformer.py]: 已打开注释

File: easycv/thirdparty/deformable_attention/src/vision.cpp (hunk 8 above)
Comment (Cathy0908) [reply easycv/thirdparty/deformable_attention/src/vision.cpp:34]: 已添加

File: easycv/models/detection3d/detectors/bevformer/bevformer_head.py (hunk 3 above)
Comment (Cathy0908) [reply easycv/models/detection3d/detectors/bevformer/bevformer_head.py:345]: inplace耗时更久

File: easycv/models/detection3d/detectors/bevformer/transformer.py (hunk 5 above)
Comment (Cathy0908) [reply easycv/models/detection3d/detectors/bevformer/transformer.py:376]: inplace耗时更久

**Review by wenmengzhou — APPROVED** [review]

**Pull request merged**